cpp/src/gandiva/decimal_ir.cc

```diff
@@ -79,13 +79,22 @@ void DecimalIR::AddGlobals(Engine* engine) {
   globalScaleMultipliers->setAlignment(LLVM_ALIGN(16));
 }
 
+#if LLVM_VERSION_MAJOR < 20
+namespace {
+inline llvm::Function* getOrInsertDeclaration(llvm::Module* M, llvm::Intrinsic::ID id,
+                                              llvm::ArrayRef<llvm::Type*> Tys) {
+  return llvm::Intrinsic::getDeclaration(M, id, Tys);
+}
+}  // namespace
+#endif
+
 // Lookup intrinsic functions
 void DecimalIR::InitializeIntrinsics() {
-  sadd_with_overflow_fn_ = llvm::Intrinsic::getDeclaration(
+  sadd_with_overflow_fn_ = getOrInsertDeclaration(
```

Comment:
I am curious on why if LLVM > 20 we don't have to specifically add `using llvm::Intrinsic::getOrInsertDeclaration`. Basically, how does it know the symbol is on that namespace if this is in declared in the `gandiva` namespace.

Reply:
I mean, it clearly is working based on the CI jobs but I am not sure why the above is not required

Reply:
Because we have `llvm::Intrinsic::sadd_with_overflow` in arguments. C++ searches a target function based on argument types. See also: https://en.cppreference.com/w/cpp/language/adl

Reply:
Thanks!